pandas/io/formats/format.py

```diff
@@ -565,315 +569,409 @@ def __init__(
-            # screen
-            if max_cols == 0 and len(self.frame.columns) > w:
-                max_cols = w
-            if max_rows == 0 and len(self.frame) > h:
-                max_rows = h
-
-        if not hasattr(self, "max_rows_adj"):
-            if max_rows:
-                if (len(self.frame) > max_rows) and self.min_rows:
-                    # if truncated, set max_rows showed to min_rows
-                    max_rows = min(self.min_rows, max_rows)
-            self.max_rows_adj = max_rows
-        if not hasattr(self, "max_cols_adj"):
-            self.max_cols_adj = max_cols
-
-        max_cols_adj = self.max_cols_adj
-        max_rows_adj = self.max_rows_adj
-
-        truncate_h = max_cols_adj and (len(self.columns) > max_cols_adj)
-        truncate_v = max_rows_adj and (len(self.frame) > max_rows_adj)
-
-        frame = self.frame
-        if truncate_h:
-            # cast here since if truncate_h is True, max_cols_adj is not None
-            max_cols_adj = cast(int, max_cols_adj)
-            if max_cols_adj == 0:
-                col_num = len(frame.columns)
-            elif max_cols_adj == 1:
-                max_cols = cast(int, max_cols)
-                frame = frame.iloc[:, :max_cols]
-                col_num = max_cols
-            else:
-                col_num = max_cols_adj // 2
-                frame = concat(
-                    (frame.iloc[:, :col_num], frame.iloc[:, -col_num:]), axis=1
-                )
-                # truncate formatter
-                if isinstance(self.formatters, (list, tuple)):
-                    truncate_fmt = self.formatters
-                    self.formatters = [
-                        *truncate_fmt[:col_num],
-                        *truncate_fmt[-col_num:],
-                    ]
-            self.tr_col_num = col_num
-        if truncate_v:
-            # cast here since if truncate_v is True, max_rows_adj is not None
-            max_rows_adj = cast(int, max_rows_adj)
-            if max_rows_adj == 1:
-                row_num = max_rows
-                frame = frame.iloc[:max_rows, :]
-            else:
-                row_num = max_rows_adj // 2
-                frame = concat((frame.iloc[:row_num, :], frame.iloc[-row_num:, :]))
-            self.tr_row_num = row_num
-        else:
-            self.tr_row_num = None
+        if self.is_truncated_horizontally:
+            self._truncate_horizontally()
 
-        self.tr_frame = frame
-        self.truncate_h = truncate_h
-        self.truncate_v = truncate_v
-        self.is_truncated = bool(self.truncate_h or self.truncate_v)
+        if self.is_truncated_vertically:
+            self._truncate_vertically()
 
-    def _to_str_columns(self) -> List[List[str]]:
-        """
-        Render a DataFrame to a list of columns (as lists of strings).
+    def _truncate_horizontally(self) -> None:
+        """Remove columns, which are not to be displayed and adjust formatters.
```

Comment:
wouldn't be against making this a non-mutating function and instead just return the new values (and set where this is called), but you did explain so ok.